source/common/router/rds_impl.cc

```diff
@@ -159,21 +173,121 @@ Router::ConfigConstSharedPtr RdsRouteConfigProviderImpl::config() {
+    const envoy::api::v2::RouteConfiguration& route_configuration,
+    Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix,
+    std::unordered_set<RdsRouteConfigProviderImpl*>& route_config_providers,
+    SubscriptionFactoryFunction factory_function)
+    : route_config_proto_(route_configuration), route_config_name_(route_configuration.name()),
+      init_target_(fmt::format("VhdsConfigSubscription {}", route_config_name_),
+                   [this]() { subscription_->start({}, *this); }),
+      scope_(factory_context.scope().createScope(stat_prefix + "vhds." + route_config_name_ + ".")),
+      stats_({ALL_VHDS_STATS(POOL_COUNTER(*scope_))}), time_source_(factory_context.timeSource()),
+      last_updated_(factory_context.timeSource().systemTime()),
+      route_config_providers_(route_config_providers) {
+  Envoy::Config::Utility::checkLocalInfo("vhds", factory_context.localInfo());
+  const auto& config_source =
+      route_configuration.vhds().config_source().api_config_source().api_type();
+  if (config_source != envoy::api::v2::core::ApiConfigSource::DELTA_GRPC) {
+    throw EnvoyException("vhds: only 'DELTA_GRPC' is supported as an api_type.");
+  }
+
+  initializeVhosts(route_config_proto_);
+  subscription_ = factory_function(
+      route_configuration.vhds().config_source(), factory_context.localInfo(),
+      factory_context.dispatcher(), factory_context.clusterManager(), factory_context.random(),
+      *scope_, "none", "envoy.api.v2.VirtualHostDiscoveryService.DeltaVirtualHosts",
+      Grpc::Common::typeUrl(envoy::api::v2::route::VirtualHost().GetDescriptor()->full_name()),
+      factory_context.api());
+}
+
+void VhdsSubscription::onConfigUpdateFailed(const EnvoyException*) {
+  // We need to allow server startup to continue, even if we have a bad
+  // config.
+  init_target_.ready();
+}
+
+void VhdsSubscription::onConfigUpdate(
+    const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+    const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+    const std::string& version_info) {
+  last_updated_ = time_source_.systemTime();
+  removeVhosts(virtual_hosts_, removed_resources);
+  updateVhosts(virtual_hosts_, added_resources);
+  rebuildRouteConfig(virtual_hosts_, route_config_proto_);
+
+  const uint64_t new_hash = MessageUtil::hash(route_config_proto_);
+  if (!config_info_ || new_hash != config_info_.value().last_config_hash_) {
+    config_info_ = {new_hash, version_info};
+    stats_.config_reload_.inc();
+    ENVOY_LOG(debug, "vhds: loading new configuration: config_name={} hash={}", route_config_name_,
+              new_hash);
+    for (auto* provider : route_config_providers_) {
+      provider->onConfigUpdate();
+    }
+  }
+
+  init_target_.ready();
+}
+
+void VhdsSubscription::initializeVhosts(
+    const envoy::api::v2::RouteConfiguration& route_configuration) {
+  if (route_configuration.virtual_hosts_size() <= 0) {
```

Comment:
Doesn't the `for` loop below basically do this?

Reply:
Hahaha. Yes, it does indeed. Fixed.